miio/tests/test_airpurifier.py

```diff
@@ -0,0 +1,186 @@
+            'set_led_b': lambda x: self._set_state("led_b", x),
+        }
+        super().__init__(args, kwargs)
+
+
+@pytest.fixture(scope="class")
+def airpurifier(request):
+    request.cls.device = DummyAirPurifier()
+    # TODO add ability to test on a real device
+
+
+@pytest.mark.usefixtures("airpurifier")
+class TestAirPurifier(TestCase):
+    def is_on(self):
+        return self.device.status().is_on
+
+    def state(self):
+        return self.device.status()
+
+    def test_on(self):
+        self.device.off()  # ensure off
+
+        start_state = self.is_on()
+        assert start_state is False
+
+        self.device.on()
+        assert self.is_on() is True
+
+    def test_off(self):
+        self.device.on()  # ensure on
+
+        assert self.is_on() is True
+        self.device.off()
+        assert self.is_on() is False
+
+    def test_status(self):
+        self.device._reset_state()
+
+        assert self.is_on() is True
+        assert self.state().aqi == self.device.start_state["aqi"]
+        assert self.state().average_aqi == self.device.start_state["average_aqi"]
+        assert self.state().temperature == self.device.start_state["temp_dec"] / 10.0
+        assert self.state().humidity == self.device.start_state["humidity"]
+        assert self.state().mode == OperationMode(self.device.start_state["mode"])
+        assert self.state().favorite_level == self.device.start_state["favorite_level"]
+        assert self.state().filter_life_remaining == self.device.start_state["filter1_life"]
+        assert self.state().filter_hours_used == self.device.start_state["f1_hour_used"]
+        assert self.state().use_time == self.device.start_state["use_time"]
+        assert self.state().motor_speed == self.device.start_state["motor1_speed"]
+        assert self.state().purify_volume == self.device.start_state["purify_volume"]
+
+        assert self.state().led == (self.device.start_state["led"] == 'on')
+        assert self.state().led_brightness == LedBrightness(self.device.start_state["led_b"])
+        assert self.state().buzzer == (self.device.start_state["buzzer"] == 'on')
+        assert self.state().child_lock == (self.device.start_state["child_lock"] == 'on')
+
+    def test_set_mode(self):
+        def mode():
+            return self.device.status().mode
+
+        self.device.set_mode(OperationMode.Silent)
+        assert mode() == OperationMode.Silent
+
+        self.device.set_mode(OperationMode.Auto)
+        assert mode() == OperationMode.Auto
+
+        self.device.set_mode(OperationMode.Favorite)
+        assert mode() == OperationMode.Favorite
+
+        self.device.set_mode(OperationMode.Idle)
+        assert mode() == OperationMode.Idle
+
+    def test_set_favorite_level(self):
+        def favorite_level():
+            return self.device.status().favorite_level
+
+        self.device.set_favorite_level(1)
+        assert favorite_level() == 1
```

Comment:
Is it known what values are allowed for favorite level? If yes, it makes probably sense to add an exception if out-of-range values are inputed, and to add a test for that also here (https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions).